<a href="https://colab.research.google.com/github/hong126-ch/CIS5450/blob/main/8_Module_2_Part_IV_Parallelism.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lecture Module 2 Part 4: Parallelism

## LinkedIn Social Analysis

We now consider extensions of processing to multi-core and multi-machine models.  This includes:

* Polars, a multi-core Pandas-style library that is seeing wide adoption due to its performance (and flexible programming model)
* Sharded execution, a technique that is the basis of Spark's execution model.


In [1]:
!wget -nc https://storage.googleapis.com/penn-cis5450/linkedin_anon.jsonl

--2025-10-06 09:44:49--  https://storage.googleapis.com/penn-cis5450/linkedin_anon.jsonl
Resolving storage.googleapis.com (storage.googleapis.com)... 64.233.180.207, 172.253.115.207, 172.253.122.207, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|64.233.180.207|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 179851696 (172M) [application/octet-stream]
Saving to: ‘linkedin_anon.jsonl’

linkedin_anon.jsonl 100%[===================>] 171.52M   216MB/s    in 0.8s    

2025-10-06 09:44:50 (216 MB/s) - ‘linkedin_anon.jsonl’ saved [179851696/179851696]



## PennGrader Setup

In [2]:
%%writefile notebook-config.yaml

grader_api_url: 'https://23whrwph9h.execute-api.us-east-1.amazonaws.com/default/Grader23'
grader_api_key: 'flfkE736fA6Z8GxMDJe2q8Kfk8UDqjsG3GVqOFOa'

Writing notebook-config.yaml


In [3]:
!pip3 install penngrader-client

In [4]:
#PLEASE ENSURE YOUR PENN-ID IS ENTERED CORRECTLY. IF NOT, THE AUTOGRADER WON'T KNOW WHO
#TO ASSIGN POINTS TO YOU IN OUR BACKEND
STUDENT_ID = 64660501 # YOUR PENN-ID GOES HERE AS AN INTEGER##PLEASE ENSURE YOUR PENN-ID IS ENTERED CORRECTLY. IF NOT, THE AUTOGRADER WON'T KNOW WHO

In [5]:
%set_env HW_ID=cis5450_25f_HW9

env: HW_ID=cis5450_25f_HW9


In [6]:
import os
from penngrader.grader import *

grader = PennGrader('notebook-config.yaml', os.environ['HW_ID'], STUDENT_ID, STUDENT_ID)

PennGrader initialized with Student ID: 64660501

Make sure this correct or we will not be able to store your grade


# Loading our Data

Now that we've seen how to do fairly complex queries over data in relations, we'll "pop back" to our big data example, which is the LinkedIn dataset.  Recall that we had a segment of the LinkedIn input file in our previous examples earlier in this module.

In [7]:
import pandas as pd
import numpy as np

# JSON parsing
import json

# Time conversions
import time

In [8]:
'''
Simple code to pull out data from JSON and load into DuckDB.
'''
linked_in = open('linkedin_anon.jsonl')
people_df = pd.read_json('linkedin_anon.jsonl', lines=True)

In [9]:
people_df

_id  \
0               moist-vodka   
1           adagio-catalyst   
2                tart-acorn   
3        objective-riesling   
4      generative-amberjack   
...                     ...   
49995         glowing-flush   
49996        grouchy-flight   
49997            dense-bell   
49998           brave-hoops   
49999       fixed-partition   

                                                    name  \
0      {'family_name': 'Post', 'given_name': 'Belvede...   
1       {'family_name': 'Watt', 'given_name': 'Brunton'}   
2      {'family_name': 'Hannay', 'given_name': 'Passe...   
3      {'family_name': 'Carnegie', 'given_name': 'Pas...   
4      {'family_name': 'Duncan', 'given_name': 'Merri...   
...                                                  ...   
49995  {'family_name': 'Kincaid', 'given_name': 'Pass...   
49996  {'family_name': 'Ogilvy', 'given_name': 'Bulli...   
49997  {'family_name': 'Macdougall', 'given_name': 'B...   
49998  {'family_name': 'Forsyth', 'given_name': 'Cadb...   
49999  {'family_name': 'Donald', 'given_name': 'Alfred'}   

                         locality  \
0                   United States   
1           Antwerp Area, Belgium   
2       San Francisco, California   
3          San Francisco Bay Area   
4             Chennai Area, India   
...                           ...   
49995        Greater Chicago Area   
49996        Greater Atlanta Area   
49997        Calgary, Canada Area   
49998      San Francisco Bay Area   
49999  Greater New York City Area   

                                                  skills  \
0      [Key Account Development, Strategic Planning, ...   
1                        [Molecular Biology, Biomarkers]   
2      [DNA, Nanotechnology, Molecular Biology, Softw...   
3                                                   None   
4      [Program Management, French, Avionics, Embedde...   
...                                                  ...   
49995  [Spanish-speaking, Cicerone-Certified Beer Ser...   
49996                                               None   
49997  [Project Management, Electrical Engineering, M...   
49998  [Corporate Social Responsibility, Public Polic...   
49999                                               None   

                                  industry  \
0                          Medical Devices   
1                          Pharmaceuticals   
2                                 Research   
3      Information Technology and Services   
4                     Aviation & Aerospace   
...                                    ...   
49995            Marketing and Advertising   
49996                   Financial Services   
49997                               Design   
49998                        Public Policy   
49999  Public Relations and Communications   

                                                 summary  \
0      SALES MANAGEMENT / BUSINESS DEVELOPMENT / PROJ...   
1      Ph.D. scientist with background in cancer rese...   
2      I am interested in inventing new methods to co...   
3      OBJECTIVE<Primary> Work on an interesting and ...   
4      Experience in Avionics Systems, Embedded Syste...   
...                                                  ...   
49995  Sales and marketing professional specializing ...   
49996  Accomplished business development manager expe...   
49997  Brad Gibson is a recognized expert in power qu...   
49998  Brad Kane's multi-faceted career in the govern...   
49999  Client Experience:CONSUMER TECH: foursquare, O...   

                                                    url  \
0               https://www.linkedin.com/in/moist-vodka   
1           https://www.linkedin.com/in/adagio-catalyst   
2                https://www.linkedin.com/in/tart-acorn   
3        https://www.linkedin.com/in/objective-riesling   
4      https://www.linkedin.com/in/generative-amberjack   
...                                                 ...   
49995         https://www.linkedin.com/in/glowing-flush   
49996        https

In [10]:
def get_nested_dict(rel, name):
  # This evaluates the string that describes the dictionary, as a dictionary
  # definition
  ret = rel.copy()
  # ret[name] = rel[name].map(lambda x: ast.literal_eval(x) if len(x) else np.NaN)
  ret = ret.dropna()
  # This joins rows on the index
  return ret.drop(columns=name).join(pd.DataFrame(ret[name].tolist()))

def get_nested_list(rel, name):
  ret = rel.copy()
  ret = ret.dropna().explode(name).dropna()
  ret = ret.join(pd.DataFrame(ret[name].tolist())).drop(columns=name).drop_duplicates()
  return ret.rename(columns={0: name})

def get_nested_list_dict(rel, name):
  ret = rel.copy()

  ret = ret.dropna().explode(name)

  exploded_pairs = pd.DataFrame(ret.apply(lambda x: {'_id': x['_id']} | x[name] if isinstance(x[name], dict) else {'_id': x['_id']}, axis=1).tolist())

  return ret.merge(exploded_pairs, on='_id').drop(columns=name)

# Take the lists, drop any blank strings
specialties_df = people_df[['_id','specilities']].explode('specilities')
specialties_df.dropna(inplace=True)
interests_df = people_df[['_id','interests']].explode('interests')
interests_df.dropna(inplace=True)

names_df = get_nested_dict(people_df[['_id','name']], 'name')

education_df = get_nested_list_dict(people_df[['_id','education']], 'education')
experience_df = get_nested_list_dict(people_df[['_id','experience']], 'experience')
skills_df = get_nested_list(people_df[['_id','skills']], 'skills')
honors_df = get_nested_list(people_df[['_id','honors']], 'honors')
events_df = get_nested_list_dict(people_df[['_id','events']], 'events')

groups_df = get_nested_dict(people_df[['_id','group']], 'group')

people_only_df = people_df.drop(columns=['name','education','group','skills','experience','honors','events','specilities','interests']).merge(names_df,on='_id')

# Multi-Core Execution of Dataframes

## Polars

Polars is a "drop-in replacement" for Pandas that supports more advanced processing, including multicore processing.

Its syntax is not exactly the same as Pandas', but close.  (In fact, it has some similarities to Apache Spark DataFrames, which we'll be seeing shortly.)

### Polars DataFrames

To take advantage of Polars you want to convert your dataframes from Pandas to Polars. (If you have base files, you can also use Polars' read_csv etc as you would from Pandas.)

In [11]:
import polars as pl

# Convert all Pandas dataframes to Polars Dataframes
people_only_pdf = pl.from_pandas(people_only_df)
education_pdf = pl.from_pandas(education_df)
experience_pdf = pl.from_pandas(experience_df)
skills_pdf = pl.from_pandas(skills_df)
groups_pdf = pl.from_pandas(groups_df)
names_pdf = pl.from_pandas(names_df)

## PL.Col

In Polars, when you are referring to a column in a DataFrame expression, this is represented by a *column object*.  For instance, you can express predicates on the values in a column when filtering, or you can request that certain columns be projected.

You can, e.g., talk about `pl.col('_id')`.  SQL-style, you can also refer to `pl.col('*')` if you are asking for all columns.

## Selection and Projection in Polars

Rather than using Pandas' `x[x[condition]]` notation, Polars has a *filter* function.   To project, you use a function called `select` (this is a bit confusing, but think SQL SELECT, not relational algebra select).

In [12]:
people_only_pdf.\
  select(pl.col('_id','family_name','given_name')).\
  filter(pl.col('family_name') <= 'Smith')

shape: (47_983, 3)
┌──────────────────────┬─────────────┬──────────────┐
│ _id                  ┆ family_name ┆ given_name   │
│ ---                  ┆ ---         ┆ ---          │
│ str                  ┆ str         ┆ str          │
╞══════════════════════╪═════════════╪══════════════╡
│ moist-vodka          ┆ Post        ┆ Belvedere    │
│ tart-acorn           ┆ Hannay      ┆ Passepartout │
│ objective-riesling   ┆ Carnegie    ┆ Passepartout │
│ generative-amberjack ┆ Duncan      ┆ Merriman     │
│ chalky-tenement      ┆ Graham      ┆ Leporello    │
│ …                    ┆ …           ┆ …            │
│ glowing-flush        ┆ Kincaid     ┆ Passepartout │
│ grouchy-flight       ┆ Ogilvy      ┆ Bullimore    │
│ dense-bell           ┆ Macdougall  ┆ Barrymore    │
│ brave-hoops          ┆ Forsyth     ┆ Cadbury      │
│ fixed-partition      ┆ Donald      ┆ Alfred       │
└──────────────────────┴─────────────┴──────────────┘

## Joins

In many settings, multi-way joins are the most expensive query operations -- partly because they "blow up the data". You saw this with our discussions about join ordering.

We should be able to see the benefits of Polars multicore processing with some join expressions. Note that we still need to keep everything in memory for Pandas/Polars, and in fact there are simple join queries (say, all people joined with experiences and education) that will cause us to run out of memory.

Let's try a simpler query that joins people with some of their other info.

### Pandas Version

In [13]:
%%time
people_only_df.merge(skills_df, on='_id').merge(education_df, on='_id').merge(groups_df, on='_id')

CPU times: user 1 s, sys: 174 ms, total: 1.18 s
Wall time: 1.21 s


_id                locality         industry  \
0       adagio-catalyst   Antwerp Area, Belgium  Pharmaceuticals   
1       adagio-catalyst   Antwerp Area, Belgium  Pharmaceuticals   
2       adagio-catalyst   Antwerp Area, Belgium  Pharmaceuticals   
3       adagio-catalyst   Antwerp Area, Belgium  Pharmaceuticals   
4       adagio-catalyst   Antwerp Area, Belgium  Pharmaceuticals   
...                 ...                     ...              ...   
368157      brave-hoops  San Francisco Bay Area    Public Policy   
368158      brave-hoops  San Francisco Bay Area    Public Policy   
368159      brave-hoops  San Francisco Bay Area    Public Policy   
368160      brave-hoops  San Francisco Bay Area    Public Policy   
368161      brave-hoops  San Francisco Bay Area    Public Policy   

                                                  summary  \
0       Ph.D. scientist with background in cancer rese...   
1       Ph.D. scientist with background in cancer rese...   
2       Ph.D. scientist with background in cancer rese...   
3       Ph.D. scientist with background in cancer rese...   
4       Ph.D. scientist with background in cancer rese...   
...                                                   ...   
368157  Brad Kane's multi-faceted career in the govern...   
368158  Brad Kane's multi-faceted career in the govern...   
368159  Brad Kane's multi-faceted career in the govern...   
368160  Brad Kane's multi-faceted career in the govern...   
368161  Brad Kane's multi-faceted career in the govern...   

                                                url  interval family_name  \
0       https://www.linkedin.com/in/adagio-catalyst      20.0        Watt   
1       https://www.linkedin.com/in/adagio-catalyst      20.0        Watt   
2       https://www.linkedin.com/in/adagio-catalyst      20.0        Watt   
3       https://www.linkedin.com/in/adagio-catalyst      20.0        Watt   
4       https://www.linkedin.com/in/adagio-catalyst      20.0        Watt   
...                                             ...       ...         ...   
368157      https://www.linkedin.com/in/brave-hoops      26.0     Forsyth   
368158      https://www.linkedin.com/in/brave-hoops      26.0     Forsyth   
368159      https://www.linkedin.com/in/brave-hoops      26.0     Forsyth   
368160      https://www.linkedin.com/in/brave-hoops      26.0     Forsyth   
368161      https://www.linkedin.com/in/brave-hoops      26.0     Forsyth   

       given_name              skills start  \
0         Brunton  Strategic Planning  2008   
1         Brunton  Strategic Planning  2007   
2         Brunton  Strategic Planning  1996   
3         Brunton  Strategic Planning  1994   
4         Brunton  Strategic Planning  1989   
...           ...                 ...   ...   
368157    Cadbury          Web Design  1978   
368158    Cadbury          Web Design   NaN   
368159    Cadbury          Web Design  1981   
368160    Cadbury          Web Design  1978   
368161    Cadbury          Web Design   NaN   

                                    major   end  \
0                               Economics  2008   
1                                     NaN  2007   
2                         Cancer genomics  2001   
3       Biology, Medicine;German Language  1995   
4                                 Biology  1994   
...                                   ...   ...   
368157            International Relations  1981   
368158                                NaN   NaN   
368159                                Law  1984   
368160            International Relations  1981   
368161                                NaN   NaN   

                                                     name  \
0          Columbia University - Columbia Business School   
1          Columbia University - Columbia Business School   
2                                   Göteborgs universitet   
3                                  Universität Regensburg   
4                                   Göteborgs universitet   
.

### Polars Version

Notice this finishes in roughly half the time as Pandas!

In [14]:
%%time
print(people_only_pdf.join(skills_pdf, on='_id').join(education_pdf, on='_id').join(groups_pdf, on='_id'))

shape: (368_162, 17)
┌────────────┬────────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬────────┐
│ _id        ┆ locality   ┆ industry  ┆ summary   ┆ … ┆ desc      ┆ degree    ┆ affilitio ┆ member │
│ ---        ┆ ---        ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ n         ┆ ---    │
│ str        ┆ str        ┆ str       ┆ str       ┆   ┆ str       ┆ str       ┆ ---       ┆ str    │
│            ┆            ┆           ┆           ┆   ┆           ┆           ┆ list[str] ┆        │
╞════════════╪════════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪════════╡
│ adagio-cat ┆ Antwerp    ┆ Pharmaceu ┆ Ph.D.     ┆ … ┆ Coursewor ┆ null      ┆ ["Big     ┆ null   │
│ alyst      ┆ Area,      ┆ ticals    ┆ scientist ┆   ┆ k "Princi ┆           ┆ Data, Low ┆        │
│            ┆ Belgium    ┆           ┆ with back ┆   ┆ pals of   ┆           ┆ Latency", ┆        │
│            ┆            ┆           ┆ groun…    ┆   ┆ Econ…     ┆   

## Laziness

In [15]:
%%time
result_pdf = people_only_pdf.join(skills_pdf, on='_id').\
  join(education_pdf, on='_id').join(groups_pdf, on='_id')
only_sf_pdf = result_pdf.filter(pl.col('locality') == 'San Francisco Bay Area')
print (len(only_sf_pdf))

10712
CPU times: user 213 ms, sys: 32.9 ms, total: 246 ms
Wall time: 167 ms


In [16]:
%%time
result_pdf = people_only_pdf.lazy().join(skills_pdf.lazy(), on='_id').join(education_pdf.lazy(), on='_id').join(groups_pdf.lazy(), on='_id').lazy()
only_sf_pdf = result_pdf.filter(pl.col('locality') == 'San Francisco Bay Area')
print (len(only_sf_pdf.collect()))

10712
CPU times: user 25.8 ms, sys: 2.79 ms, total: 28.6 ms
Wall time: 19.5 ms


In [17]:
results = only_sf_pdf.explain()

for line in results.split('\n'):
  print(line)

INNER JOIN:
LEFT PLAN ON: [col("_id")]
  INNER JOIN:
  LEFT PLAN ON: [col("_id")]
    INNER JOIN:
    LEFT PLAN ON: [col("_id")]
      FILTER [(col("locality")) == (String(San Francisco Bay Area))] FROM
        DF ["_id", "locality", "industry", "summary", ...]; PROJECT */8 COLUMNS
    RIGHT PLAN ON: [col("_id")]
      DF ["_id", "skills"]; PROJECT */2 COLUMNS
    END INNER JOIN
  RIGHT PLAN ON: [col("_id")]
    DF ["_id", "start", "major", "end", ...]; PROJECT */7 COLUMNS
  END INNER JOIN
RIGHT PLAN ON: [col("_id")]
  DF ["_id", "affilition", "member"]; PROJECT */3 COLUMNS
END INNER JOIN


# Understanding Sharded Execution

Let's now look (in simulation) at how sharding works!  This is the basis of cluster-based processing.

## A Data Graph in Relations

To do this we'll adapt our LinkedIn data to a *graph* scenario, with people and companies.

In [18]:
persons_df = people_only_df[['_id','family_name','given_name']]

persons_df

_id family_name    given_name
0               moist-vodka        Post     Belvedere
1           adagio-catalyst        Watt       Brunton
2                tart-acorn      Hannay  Passepartout
3        objective-riesling    Carnegie  Passepartout
4      generative-amberjack      Duncan      Merriman
...                     ...         ...           ...
51817         glowing-flush     Kincaid  Passepartout
51818        grouchy-flight      Ogilvy     Bullimore
51819            dense-bell  Macdougall     Barrymore
51820           brave-hoops     Forsyth       Cadbury
51821       fixed-partition      Donald        Alfred

[51822 rows x 3 columns]

In [19]:
experience_df

_id                                     org  \
0        adagio-catalyst                     Johnson and Johnson   
1        adagio-catalyst          Albert Einstein Medical Center   
2        adagio-catalyst                     Columbia University   
3        adagio-catalyst  Memorial Sloan Kettering Cancer Center   
4        adagio-catalyst         Sahlgrenska University Hospital   
...                  ...                                     ...   
1761792  fixed-partition         Four Seasons Hotels and Resorts   
1761793  fixed-partition                    M Booth & Associates   
1761794  fixed-partition                         Hill & Knowlton   
1761795  fixed-partition               Richmond Public Relations   
1761796  fixed-partition         Four Seasons Hotels and Resorts   

                                           title      end           start  \
0          Senior Scientist, Oncology Biomarkers  Present   November 2009   
1        Associate at Dept of Molecular Genetics      NaN  September 2008   
2                   Associate Research Scientist      NaN     August 2006   
3                  Post Doctoral Research Fellow      NaN    January 2003   
4                             Research Scientist      NaN   November 2001   
...                                          ...      ...             ...   
1761792                 Communications Assistant      NaN            1995   
1761793          Senior Vice President, Director  Present      March 2002   
1761794                       Account Supervisor      NaN  September 1999   
1761795                        Account Executive      NaN            1996   
1761796                 Communications Assistant      NaN            1995   

                                                      desc  
0        Biomarker Leader for compounds in clinical dev...  
1                             Single Cell Gene expression.  
2        Work on peptide to restore wt p53 function in ...  
3                Molecular profiling of colorectal cancer.  
4        Cancer Research at Dept of Surgery.Molecular p...  
...                                                    ...  
1761792                                                NaN  
1761793                   Director, Consumer Tech & Travel  
1761794                                                NaN  
1761795                                                NaN  
1761796                                                NaN  

[1761797 rows x 6 columns]

Companies are also nodes.  We'll auto-assign IDs for these.

In [20]:
companies_df = experience_df[['org']].drop_duplicates()
companies_df['company_id'] = companies_df.index
companies_df

org  company_id
0                                      Johnson and Johnson           0
1                           Albert Einstein Medical Center           1
2                                      Columbia University           2
3                   Memorial Sloan Kettering Cancer Center           3
4                          Sahlgrenska University Hospital           4
...                                                    ...         ...
1761646       President's Task Force on Health Care Reform     1761646
1761647  Energy and Commerce Subcommittee on Commerce, ...     1761647
1761648                      Congresswoman Cardiss Collins     1761648
1761781                               M Booth & Associates     1761781
1761783                          Richmond Public Relations     1761783

[145703 rows x 2 columns]

Finally, here are some edges.  We preserve the relationships between people and their experiences with organizations.

In [21]:
worked_for = experience_df[['_id', 'org', 'title']].copy().merge(companies_df.set_index('org'), on='org').drop_duplicates()
worked_for = worked_for.drop(columns=['org'])
worked_for

_id                                              title  \
0        adagio-catalyst              Senior Scientist, Oncology Biomarkers   
1        adagio-catalyst            Associate at Dept of Molecular Genetics   
2        adagio-catalyst                       Associate Research Scientist   
3        adagio-catalyst                      Post Doctoral Research Fellow   
4        adagio-catalyst                                 Research Scientist   
...                  ...                                                ...   
1761648      brave-hoops  Legislative Counsel: U.S. House of Representat...   
1761781  fixed-partition                    Senior Vice President, Director   
1761782  fixed-partition                                 Account Supervisor   
1761783  fixed-partition                                  Account Executive   
1761784  fixed-partition                           Communications Assistant   

         company_id  
0                 0  
1                 1  
2                 2  
3                 3  
4                 4  
...             ...  
1761648     1761648  
1761781     1761781  
1761782      146464  
1761783     1761783  
1761784      186668  

[230440 rows x 3 columns]

### Data

In [22]:
print('persons_df:')
display(persons_df)
print('companies_df:')
display(companies_df)
print('worked_df:')
display(worked_for)

persons_df:


_id family_name    given_name
0               moist-vodka        Post     Belvedere
1           adagio-catalyst        Watt       Brunton
2                tart-acorn      Hannay  Passepartout
3        objective-riesling    Carnegie  Passepartout
4      generative-amberjack      Duncan      Merriman
...                     ...         ...           ...
51817         glowing-flush     Kincaid  Passepartout
51818        grouchy-flight      Ogilvy     Bullimore
51819            dense-bell  Macdougall     Barrymore
51820           brave-hoops     Forsyth       Cadbury
51821       fixed-partition      Donald        Alfred

[51822 rows x 3 columns]

companies_df:


org  company_id
0                                      Johnson and Johnson           0
1                           Albert Einstein Medical Center           1
2                                      Columbia University           2
3                   Memorial Sloan Kettering Cancer Center           3
4                          Sahlgrenska University Hospital           4
...                                                    ...         ...
1761646       President's Task Force on Health Care Reform     1761646
1761647  Energy and Commerce Subcommittee on Commerce, ...     1761647
1761648                      Congresswoman Cardiss Collins     1761648
1761781                               M Booth & Associates     1761781
1761783                          Richmond Public Relations     1761783

[145703 rows x 2 columns]

worked_df:


_id                                              title  \
0        adagio-catalyst              Senior Scientist, Oncology Biomarkers   
1        adagio-catalyst            Associate at Dept of Molecular Genetics   
2        adagio-catalyst                       Associate Research Scientist   
3        adagio-catalyst                      Post Doctoral Research Fellow   
4        adagio-catalyst                                 Research Scientist   
...                  ...                                                ...   
1761648      brave-hoops  Legislative Counsel: U.S. House of Representat...   
1761781  fixed-partition                    Senior Vice President, Director   
1761782  fixed-partition                                 Account Supervisor   
1761783  fixed-partition                                  Account Executive   
1761784  fixed-partition                           Communications Assistant   

         company_id  
0                 0  
1                 1  
2                 2  
3                 3  
4                 4  
...             ...  
1761648     1761648  
1761781     1761781  
1761782      146464  
1761783     1761783  
1761784      186668  

[230440 rows x 3 columns]

### Implementing Sharded Computations - Our Operators

Rather than look at sharding in Spark, we'll instead simulate having 4 machines, each of which has access to a shard of each table.

To illustrate the basic concepts we'll create our own basic implementations of Relational Algebra over sharded data -- as well as our own `repartition` etc.

In [23]:
nodes = 4

def get_hash(x):
  '''
  Hash code for data, used to compute a shard
  '''
  if isinstance(x, int):
    return x
  elif isinstance(x, str):
    return x.__hash__()
  elif pd.isna(x):
    return 'null'.__hash__()
  else:
    raise RuntimeError('Cannot hash {}'.format(type(x)))

def get_shard(sharded_df: list[pd.DataFrame], index: int):
  '''
  Retrieves the sub-dataframe corresponding to a shard
  '''
  return sharded_df[index]

def repartition(df_or_shards, key: str, shards: int = nodes):
  '''
  Take a dataframe or a list of sharded sub-dataframes.
  Repartition them on the key, based on the number of nodes.
  Return the new list of sub-dataframes.
  '''
  data = df_or_shards
  if isinstance(df_or_shards, list):
    data = pd.concat(df_or_shards)

  resharded_df = []
  for shard in range(0, shards):
    resharded_df.append(data[data[key].apply(lambda x: get_hash(x) % shards == shard)])

  return resharded_df

def rename_shards(sharded_df: list[pd.DataFrame], renamer: dict):
  '''
  Simple relational algebra operator:
  Rename the columns of a list of sharded dataframes
  '''
  renamed_df = []
  for shard in range(0, len(sharded_df)):
    renamed_df.append(sharded_df[shard].rename(columns=renamer))

  return renamed_df

def filter_shards(sharded_df: list[pd.DataFrame], filter_fn):
  '''
  Simple relational algebra operator:
  Filter a list of sharded dataframes with a predicate
  '''
  filtered_df = []
  for shard in range(0, len(sharded_df)):
    filtered_df.append(
        sharded_df[shard][
            sharded_df[shard].apply(filter_fn, axis=1)])

  return filtered_df


def join_shards(sharded_df1: list[pd.DataFrame], sharded_df2: list[pd.DataFrame],
                left_key, right_key):
  '''
  Simple relational algebra operator:
  Join two lists of sharded dataframes
  '''
  ret_result = []

  for shard in range(0, len(sharded_df1)):
    ret_result.append(sharded_df1[shard].merge(sharded_df2[shard], left_on=left_key, right_on=right_key))

  return ret_result

def join_shard_with_broadcast(sharded_df1: list[pd.DataFrame], sharded_df2: pd.DataFrame,
                left_key, right_key):
  '''
  Simple relational algebra operator:
  Join a list of sharded dataframes with a single dataframe
  that conceptually gets "broadcast" to all of the shards
  '''
  ret_result = []

  for shard in range(0, len(sharded_df1)):
    ret_result.append(sharded_df1[shard].merge(sharded_df2, left_on=left_key, right_on=right_key))

  return ret_result

## An Example Query

Suppose we simply want to ask this query:

`(persons_df) -[worked_for]--> (companies_df)`

## Shard the Relations!

We'll create versions of this:
* `sharded_persons` (by _id)
* `sharded_companies` (by company)
* Two shards of `worked_for`:
  - `sharded_by_user`
  - `sharded_by_company`

In [24]:
# Generally, for a table like persons, we would
# shard by default on the key (_id).  We can
# see each shard here.

sharded_persons = repartition(persons_df, '_id')

for p in sharded_persons:
  display(p)

_id family_name    given_name
0             moist-vodka        Post     Belvedere
1         adagio-catalyst        Watt       Brunton
3      objective-riesling    Carnegie  Passepartout
6         chalky-tenement      Graham     Leporello
16          plain-torpedo    Hamilton       Brunton
...                   ...         ...           ...
51803  orthogonal-liqueur     Maclean       Cadbury
51813         joint-atoll  Macalister       Cadbury
51817       glowing-flush     Kincaid  Passepartout
51818      grouchy-flight      Ogilvy     Bullimore
51819          dense-bell  Macdougall     Barrymore

[12827 rows x 3 columns]

_id family_name    given_name
2                tart-acorn      Hannay  Passepartout
4      generative-amberjack      Duncan      Merriman
5             salty-section        Watt        Figaro
8           arctic-investor       Rollo      Merriman
9                rich-laser    Spalding        Figaro
...                     ...         ...           ...
51807      sleek-sopapillas    Wheatley     Barrymore
51808       resolving-image       Agnew       Jenkins
51816          tense-spline  Cunningham      Merriman
51820           brave-hoops     Forsyth       Cadbury
51821       fixed-partition      Donald        Alfred

[12926 rows x 3 columns]

_id family_name    given_name
11     afraid-collision      Hunter      Merriman
12           oily-sport     Burnett      Merriman
13     forgiving-desert        Haig     Belvedere
15        medium-marlin    Primrose     Belvedere
24          fancy-notch   Macmillan     Barrymore
...                 ...         ...           ...
51805     radial-ribbon    Sinclair      Merriman
51806   direct-scenario        Boyd       Brunton
51809       legato-rook        Keen     Belvedere
51810        late-cadet    Mcmullen        Jeeves
51811        late-cadet  Sutherland  Passepartout

[13043 rows x 3 columns]

_id family_name    given_name
7      sensitive-estuary     Macneil      Merriman
17       cheerful-broker     Barclay       Cadbury
18            coped-wash    Buchanan        Figaro
19             hard-node       Allen       Brunton
22         tart-velocity  Sutherland         Poole
...                  ...         ...           ...
51802    pizzicato-burst       Lough        Jeeves
51804   espressivo-media      Dundas       Brunton
51812       flat-leading       Lewis       Cadbury
51814   natural-customer   Charteris  Passepartout
51815    resultant-coach     Kincaid     Bertuccio

[13026 rows x 3 columns]

In [25]:
# Similarly with companies... we shard by company_id
sharded_companies = repartition(companies_df, 'company_id')

for c in sharded_companies:
  display(c)

org  company_id
0                       Johnson and Johnson           0
4           Sahlgrenska University Hospital           4
32            <Software Consulting Company>          32
40                      <Employment Agency>          40
44                                <Biotech>          44
...                                     ...         ...
1761492           Goose Island Beer Company     1761492
1761576                           WestWayne     1761576
1761624                        The EA Group     1761624
1761644  CA Manufacturing Technology Center     1761644
1761648       Congresswoman Cardiss Collins     1761648

[36336 rows x 2 columns]

org  company_id
1                           Albert Einstein Medical Center           1
25                                                    UCSF          25
29                             <Online Recruiting Company>          29
37                               <Psychology Organization>          37
41                           <Biometrics Hardware Company>          41
...                                                    ...         ...
1761621                                             DIALOG     1761621
1761637                              The Bipartisan Bridge     1761637
1761641  Nielsen Media's Independent Task Force on Tele...     1761641
1761645               California State Controller’s Office     1761645
1761781                               M Booth & Associates     1761781

[36362 rows x 2 columns]

org  company_id
2                                      Columbia University           2
26       Wyss Institute for Biologically Inspired Engin...          26
30                               <Medical Testing Company>          30
34                       <Temperature Sensor Manufacturer>          34
42                         <Product Manufacturing Company>          42
...                                                    ...         ...
1761574                                    Celulosa Arauco     1761574
1761578                                     Bayer Advanced     1761578
1761638  International Commission on Workforce Development     1761638
1761642  State of California Business, Transportation, ...     1761642
1761646       President's Task Force on Health Care Reform     1761646

[36519 rows x 2 columns]

org  company_id
3                   Memorial Sloan Kettering Cancer Center           3
31                                       000Montgomery.Com          31
35                                   <Advertising Company>          35
39                          <National Recruitment Network>          39
43                                    <Emergency Services>          43
...                                                    ...         ...
1761575                                  House Parts, Inc.     1761575
1761623                              Current Thinking Inc.     1761623
1761639  United Nations, Global Alliance on ICT (inform...     1761639
1761647  Energy and Commerce Subcommittee on Commerce, ...     1761647
1761783                          Richmond Public Relations     1761783

[36486 rows x 2 columns]

Edges are inherently trickier to shard: we can shard by the *source* or by the *destination* but can't simultaneously do so for a single table, by both.  We can, of course, create two dataframes - one sharded by each.

In [26]:
# Edge sharded by user ID
sharded_by_user = repartition(worked_for, '_id')

for edge in sharded_by_user:
  display(edge)

_id                                    title  company_id
0        adagio-catalyst    Senior Scientist, Oncology Biomarkers           0
1        adagio-catalyst  Associate at Dept of Molecular Genetics           1
2        adagio-catalyst             Associate Research Scientist           2
3        adagio-catalyst            Post Doctoral Research Fellow           3
4        adagio-catalyst                       Research Scientist           4
...                  ...                                      ...         ...
1761578   grouchy-flight                                   Intern     1761578
1761621       dense-bell                                Principal     1761621
1761622       dense-bell                          Project Manager      913976
1761623       dense-bell                         Managing Partner     1761623
1761624       dense-bell                   Manager of Engineering     1761624

[57272 rows x 3 columns]

_id                                              title  \
25            tart-acorn                                Assistant Professor   
26            tart-acorn                      Technology Development Fellow   
870        salty-section      Founding Partner and Client Services Director   
871        salty-section                                   Managing Partner   
872        salty-section                            Senior Account Director   
...                  ...                                                ...   
1761648      brave-hoops  Legislative Counsel: U.S. House of Representat...   
1761781  fixed-partition                    Senior Vice President, Director   
1761782  fixed-partition                                 Account Supervisor   
1761783  fixed-partition                                  Account Executive   
1761784  fixed-partition                           Communications Assistant   

         company_id  
25               25  
26               26  
870             870  
871             871  
872             872  
...             ...  
1761648     1761648  
1761781     1761781  
1761782      146464  
1761783     1761783  
1761784      186668  

[57401 rows x 3 columns]

_id                                              title  \
1080     afraid-collision                                       Zaakvoerster   
1081     afraid-collision                                                 HR   
1082     afraid-collision  Branch Manager Kapellen at Adecco - Maritime/I...   
1083     afraid-collision                                     Branch Manager   
1084     afraid-collision                      Branch Manager Schoten - Lier   
...                   ...                                                ...   
1761101   direct-scenario                                   Summer Associate   
1761102   direct-scenario                         Investment Banking Analyst   
1761170       legato-rook                      Member of the Technical Staff   
1761171       legato-rook                                    Chief Architect   
1761172       legato-rook                                            Founder   

         company_id  
1080           1080  
1081           1081  
1082           1082  
1083           1082  
1084           1084  
...             ...  
1761101      327514  
1761102       32842  
1761170        5151  
1761171     1761171  
1761172     1761172  

[57667 rows x 3 columns]

_id                                              title  \
979      sensitive-estuary                      Sales Manager - Security Team   
980      sensitive-estuary                    Information Security Consultant   
981      sensitive-estuary                       IT Security Sales Consultant   
982      sensitive-estuary                                      Sales Manager   
983      sensitive-estuary                          Corporate Account Manager   
...                    ...                                                ...   
1761457    resultant-coach  Director - Head of CDO Management, Asset Backe...   
1761458    resultant-coach                       Chairman, Board of Directors   
1761459    resultant-coach                                          Associate   
1761460    resultant-coach                                          Associate   
1761461    resultant-coach                                          Associate   

         company_id  
979             979  
980             979  
981             981  
982             982  
983             983  
...             ...  
1761457      289179  
1761458     1761458  
1761459     1761459  
1761460     1761460  
1761461      310202  

[58100 rows x 3 columns]

Here's the other version...  Sharded by the company ID.

In [27]:
sharded_by_companies = repartition(worked_for, 'company_id')

for edge in sharded_by_companies:
  display(edge)

_id  \
0           adagio-catalyst   
4           adagio-catalyst   
32       objective-riesling   
33       objective-riesling   
40       objective-riesling   
...                     ...   
1761624          dense-bell   
1761644         brave-hoops   
1761648         brave-hoops   
1761782     fixed-partition   
1761784     fixed-partition   

                                                     title  company_id  
0                    Senior Scientist, Oncology Biomarkers           0  
4                                       Research Scientist           4  
32                     ASP.NET / C# / SQL 2005 / XML / XSD          32  
33                                             PHP / MySQL          32  
40                                       ASP / CRM / HRXML          40  
...                                                    ...         ...  
1761624                             Manager of Engineering     1761624  
1761644               Vice President, Government Relations     1761644  
1761648  Legislative Counsel: U.S. House of Representat...     1761648  
1761782                                 Account Supervisor      146464  
1761784                           Communications Assistant      186668  

[58333 rows x 3 columns]

_id  \
1           adagio-catalyst   
25               tart-acorn   
29       objective-riesling   
37       objective-riesling   
41       objective-riesling   
...                     ...   
1761637         brave-hoops   
1761640         brave-hoops   
1761641         brave-hoops   
1761645         brave-hoops   
1761781     fixed-partition   

                                                     title  company_id  
1                  Associate at Dept of Molecular Genetics           1  
25                                     Assistant Professor          25  
29                                                    Ning          29  
37                                               MS Access          37  
41       ASP.NET / C# / SQL 2005 / IIS 6.0 / Clients Cu...          41  
...                                                    ...         ...  
1761637                       Executive Director / Founder     1761637  
1761640  Director of Strategic Initiatives & Internatio...     1625009  
1761641                                  Task Force Member     1761641  
1761645                     Deputy Controller, Legislation     1761645  
1761781                    Senior Vice President, Director     1761781  

[56092 rows x 3 columns]

_id  \
2           adagio-catalyst   
26               tart-acorn   
30       objective-riesling   
34       objective-riesling   
42       objective-riesling   
...                     ...   
1761577      grouchy-flight   
1761578      grouchy-flight   
1761638         brave-hoops   
1761642         brave-hoops   
1761646         brave-hoops   

                                                     title  company_id  
2                             Associate Research Scientist           2  
26                           Technology Development Fellow          26  
30       .Net 4.0 / C# / VS 2010 / Silverlight 4.0 / Bl...          30  
34                                              VB 6 / SQL          34  
42                                               EDI setup          42  
...                                                    ...         ...  
1761577                          Account Management Intern       56538  
1761578                                             Intern     1761578  
1761638                            Chief Executive Officer     1761638  
1761642        Deputy Secretary for Information Technology     1761642  
1761646                                  Task Force Member     1761646  

[57430 rows x 3 columns]

_id                                       title  \
3           adagio-catalyst               Post Doctoral Research Fellow   
31       objective-riesling                                       Owner   
35       objective-riesling                 ASP.NET / WSDL / JavaScript   
36       objective-riesling                     ASP.NET / C# / SQL 2005   
38       objective-riesling                C# / MS Vista / Ant Profiler   
...                     ...                                         ...   
1761623          dense-bell                            Managing Partner   
1761639         brave-hoops                     Strategy Council Member   
1761643         brave-hoops  Executive Management & Consulting Services   
1761647         brave-hoops      Counsel: U.S. House of Representatives   
1761783     fixed-partition                           Account Executive   

         company_id  
3                 3  
31               31  
35               35  
36               35  
38               35  
...             ...  
1761623     1761623  
1761639     1761639  
1761643      110295  
1761647     1761647  
1761783     1761783  

[58585 rows x 3 columns]

## Let's Do Some Computation!

### Single Join: People -> works_for

To do this, we need to be careful to join only between tables that are sharded by the join key (_id).

In [28]:
# sharded_persons is sharded by _id
# sharded_by_user is worked_for sharded by _id
for table in join_shards(sharded_persons, sharded_by_user, '_id', '_id'):
  display(table)

# Show the final results of the join
display(pd.concat(join_shards(sharded_persons, sharded_by_user, '_id', '_id')))

_id family_name given_name  \
0      adagio-catalyst        Watt    Brunton   
1      adagio-catalyst        Watt    Brunton   
2      adagio-catalyst        Watt    Brunton   
3      adagio-catalyst        Watt    Brunton   
4      adagio-catalyst        Watt    Brunton   
...                ...         ...        ...   
63626   grouchy-flight      Ogilvy  Bullimore   
63627       dense-bell  Macdougall  Barrymore   
63628       dense-bell  Macdougall  Barrymore   
63629       dense-bell  Macdougall  Barrymore   
63630       dense-bell  Macdougall  Barrymore   

                                         title  company_id  
0        Senior Scientist, Oncology Biomarkers           0  
1      Associate at Dept of Molecular Genetics           1  
2                 Associate Research Scientist           2  
3                Post Doctoral Research Fellow           3  
4                           Research Scientist           4  
...                                        ...         ...  
63626                                   Intern     1761578  
63627                                Principal     1761621  
63628                          Project Manager      913976  
63629                         Managing Partner     1761623  
63630                   Manager of Engineering     1761624  

[63631 rows x 5 columns]

_id family_name    given_name  \
0           tart-acorn      Hannay  Passepartout   
1           tart-acorn      Hannay  Passepartout   
2        salty-section        Watt        Figaro   
3        salty-section        Watt        Figaro   
4        salty-section        Watt        Figaro   
...                ...         ...           ...   
63937      brave-hoops     Forsyth       Cadbury   
63938  fixed-partition      Donald        Alfred   
63939  fixed-partition      Donald        Alfred   
63940  fixed-partition      Donald        Alfred   
63941  fixed-partition      Donald        Alfred   

                                                   title  company_id  
0                                    Assistant Professor          25  
1                          Technology Development Fellow          26  
2          Founding Partner and Client Services Director         870  
3                                       Managing Partner         871  
4                                Senior Account Director         872  
...                                                  ...         ...  
63937  Legislative Counsel: U.S. House of Representat...     1761648  
63938                    Senior Vice President, Director     1761781  
63939                                 Account Supervisor      146464  
63940                                  Account Executive     1761783  
63941                           Communications Assistant      186668  

[63942 rows x 5 columns]

_id family_name    given_name  \
0      afraid-collision      Hunter      Merriman   
1      afraid-collision      Hunter      Merriman   
2      afraid-collision      Hunter      Merriman   
3      afraid-collision      Hunter      Merriman   
4      afraid-collision      Hunter      Merriman   
...                 ...         ...           ...   
64050        late-cadet  Sutherland  Passepartout   
64051        late-cadet  Sutherland  Passepartout   
64052        late-cadet  Sutherland  Passepartout   
64053        late-cadet  Sutherland  Passepartout   
64054        late-cadet  Sutherland  Passepartout   

                                                   title  company_id  
0                                           Zaakvoerster        1080  
1                                                     HR        1081  
2      Branch Manager Kapellen at Adecco - Maritime/I...        1082  
3                                         Branch Manager        1082  
4                          Branch Manager Schoten - Lier        1084  
...                                                  ...         ...  
64050                                    Project Manager       81256  
64051                                        IT Director      995714  
64052                             SR IS Business Analyst      890751  
64053                                            Manager       13999  
64054                                   Business Analyst      995717  

[64055 rows x 5 columns]

_id family_name given_name  \
0      sensitive-estuary     Macneil   Merriman   
1      sensitive-estuary     Macneil   Merriman   
2      sensitive-estuary     Macneil   Merriman   
3      sensitive-estuary     Macneil   Merriman   
4      sensitive-estuary     Macneil   Merriman   
...                  ...         ...        ...   
65190    resultant-coach     Kincaid  Bertuccio   
65191    resultant-coach     Kincaid  Bertuccio   
65192    resultant-coach     Kincaid  Bertuccio   
65193    resultant-coach     Kincaid  Bertuccio   
65194    resultant-coach     Kincaid  Bertuccio   

                                                   title  company_id  
0                          Sales Manager - Security Team         979  
1                        Information Security Consultant         979  
2                           IT Security Sales Consultant         981  
3                                          Sales Manager         982  
4                              Corporate Account Manager         983  
...                                                  ...         ...  
65190  Director - Head of CDO Management, Asset Backe...      289179  
65191                       Chairman, Board of Directors     1761458  
65192                                          Associate     1761459  
65193                                          Associate     1761460  
65194                                          Associate      310202  

[65195 rows x 5 columns]

_id family_name given_name  \
0      adagio-catalyst        Watt    Brunton   
1      adagio-catalyst        Watt    Brunton   
2      adagio-catalyst        Watt    Brunton   
3      adagio-catalyst        Watt    Brunton   
4      adagio-catalyst        Watt    Brunton   
...                ...         ...        ...   
65190  resultant-coach     Kincaid  Bertuccio   
65191  resultant-coach     Kincaid  Bertuccio   
65192  resultant-coach     Kincaid  Bertuccio   
65193  resultant-coach     Kincaid  Bertuccio   
65194  resultant-coach     Kincaid  Bertuccio   

                                                   title  company_id  
0                  Senior Scientist, Oncology Biomarkers           0  
1                Associate at Dept of Molecular Genetics           1  
2                           Associate Research Scientist           2  
3                          Post Doctoral Research Fellow           3  
4                                     Research Scientist           4  
...                                                  ...         ...  
65190  Director - Head of CDO Management, Asset Backe...      289179  
65191                       Chairman, Board of Directors     1761458  
65192                                          Associate     1761459  
65193                                          Associate     1761460  
65194                                          Associate      310202  

[256823 rows x 5 columns]

So, 256823 rows!

We can prove the above is correct by looking at the standard, unsharded result.

In [29]:
# Traditional Pandas query, check the count and results
persons_df.merge(worked_for, left_on='_id', right_on='_id')

_id family_name given_name  \
0       adagio-catalyst        Watt    Brunton   
1       adagio-catalyst        Watt    Brunton   
2       adagio-catalyst        Watt    Brunton   
3       adagio-catalyst        Watt    Brunton   
4       adagio-catalyst        Watt    Brunton   
...                 ...         ...        ...   
256818      brave-hoops     Forsyth    Cadbury   
256819  fixed-partition      Donald     Alfred   
256820  fixed-partition      Donald     Alfred   
256821  fixed-partition      Donald     Alfred   
256822  fixed-partition      Donald     Alfred   

                                                    title  company_id  
0                   Senior Scientist, Oncology Biomarkers           0  
1                 Associate at Dept of Molecular Genetics           1  
2                            Associate Research Scientist           2  
3                           Post Doctoral Research Fellow           3  
4                                      Research Scientist           4  
...                                                   ...         ...  
256818  Legislative Counsel: U.S. House of Representat...     1761648  
256819                    Senior Vice President, Director     1761781  
256820                                 Account Supervisor      146464  
256821                                  Account Executive     1761783  
256822                           Communications Assistant      186668  

[256823 rows x 5 columns]

### What If We Don't Watch Our Sharding?

What if we don't align the shard columns and join keys? Let's see...

In [30]:
# sharded_persons is sharded by _id
# sharded_by_companies is worked_for sharded by the company

# Output the result of the join without proper sharding
pd.concat(join_shards(sharded_persons, sharded_by_companies, '_id', '_id'))

_id family_name    given_name  \
0         adagio-catalyst        Watt       Brunton   
1         adagio-catalyst        Watt       Brunton   
2      objective-riesling    Carnegie  Passepartout   
3      objective-riesling    Carnegie  Passepartout   
4      objective-riesling    Carnegie  Passepartout   
...                   ...         ...           ...   
16544    espressivo-media      Dundas       Brunton   
16545    espressivo-media      Dundas       Brunton   
16546    natural-customer   Charteris  Passepartout   
16547     resultant-coach     Kincaid     Bertuccio   
16548     resultant-coach     Kincaid     Bertuccio   

                                                   title  company_id  
0                  Senior Scientist, Oncology Biomarkers           0  
1                                     Research Scientist           4  
2                    ASP.NET / C# / SQL 2005 / XML / XSD          32  
3                                            PHP / MySQL          32  
4                                      ASP / CRM / HRXML          40  
...                                                  ...         ...  
16544               Senior Director, Product Development     1760983  
16545                                          Associate       20167  
16546                    District Team Leader - Oklahoma     1761443  
16547  Director - Head of CDO Management, Asset Backe...      289179  
16548                                          Associate     1761459  

[64158 rows x 5 columns]

What happens if we *broadcast* one of the relations?

We would generally only do this with a small table.

In [31]:
pd.concat(join_shard_with_broadcast(sharded_persons, worked_for, '_id', '_id'))

_id family_name given_name  \
0      adagio-catalyst        Watt    Brunton   
1      adagio-catalyst        Watt    Brunton   
2      adagio-catalyst        Watt    Brunton   
3      adagio-catalyst        Watt    Brunton   
4      adagio-catalyst        Watt    Brunton   
...                ...         ...        ...   
65190  resultant-coach     Kincaid  Bertuccio   
65191  resultant-coach     Kincaid  Bertuccio   
65192  resultant-coach     Kincaid  Bertuccio   
65193  resultant-coach     Kincaid  Bertuccio   
65194  resultant-coach     Kincaid  Bertuccio   

                                                   title  company_id  
0                  Senior Scientist, Oncology Biomarkers           0  
1                Associate at Dept of Molecular Genetics           1  
2                           Associate Research Scientist           2  
3                          Post Doctoral Research Fellow           3  
4                                     Research Scientist           4  
...                                                  ...         ...  
65190  Director - Head of CDO Management, Asset Backe...      289179  
65191                       Chairman, Board of Directors     1761458  
65192                                          Associate     1761459  
65193                                          Associate     1761460  
65194                                          Associate      310202  

[256823 rows x 5 columns]

## Now Let's Do a Full Connection

OK, we saw one join.  If we want to traverse our network from people to companies, we need two joins.

Here's the Pandas query we want to do in a sharded model...

In [32]:
persons_df.merge(worked_for, left_on='_id', right_on='_id').merge(companies_df, left_on='company_id', right_on='company_id')

_id family_name given_name  \
0       adagio-catalyst        Watt    Brunton   
1       adagio-catalyst        Watt    Brunton   
2       adagio-catalyst        Watt    Brunton   
3       adagio-catalyst        Watt    Brunton   
4       adagio-catalyst        Watt    Brunton   
...                 ...         ...        ...   
256818      brave-hoops     Forsyth    Cadbury   
256819  fixed-partition      Donald     Alfred   
256820  fixed-partition      Donald     Alfred   
256821  fixed-partition      Donald     Alfred   
256822  fixed-partition      Donald     Alfred   

                                                    title  company_id  \
0                   Senior Scientist, Oncology Biomarkers           0   
1                 Associate at Dept of Molecular Genetics           1   
2                            Associate Research Scientist           2   
3                           Post Doctoral Research Fellow           3   
4                                      Research Scientist           4   
...                                                   ...         ...   
256818  Legislative Counsel: U.S. House of Representat...     1761648   
256819                    Senior Vice President, Director     1761781   
256820                                 Account Supervisor      146464   
256821                                  Account Executive     1761783   
256822                           Communications Assistant      186668   

                                           org  
0                          Johnson and Johnson  
1               Albert Einstein Medical Center  
2                          Columbia University  
3       Memorial Sloan Kettering Cancer Center  
4              Sahlgrenska University Hospital  
...                                        ...  
256818           Congresswoman Cardiss Collins  
256819                    M Booth & Associates  
256820                         Hill & Knowlton  
256821               Richmond Public Relations  
256822         Four Seasons Hotels and Resorts  

[256823 rows x 6 columns]

To join with `companies_df` we need to repartition our intermediate result on the company.

In [33]:
# This is the join from above
persons_work_for = join_shards(sharded_persons, sharded_by_user, '_id', '_id')

# persons_work_for is sharded by _id
# sharded_companies is sharded by company
# to join them, we need to find a common sharding key -- the company

# SO: we repartition persons_work_for by company
persons_companies = join_shards(repartition(persons_work_for, 'company_id'), sharded_companies, 'company_id', 'company_id')

pd.concat(
persons_companies
)

_id family_name    given_name  \
0         adagio-catalyst        Watt       Brunton   
1         adagio-catalyst        Watt       Brunton   
2      objective-riesling    Carnegie  Passepartout   
3      objective-riesling    Carnegie  Passepartout   
4      objective-riesling    Carnegie  Passepartout   
...                   ...         ...           ...   
65338    espressivo-media      Dundas       Brunton   
65339    espressivo-media      Dundas       Brunton   
65340    natural-customer   Charteris  Passepartout   
65341     resultant-coach     Kincaid     Bertuccio   
65342     resultant-coach     Kincaid     Bertuccio   

                                                   title  company_id  \
0                  Senior Scientist, Oncology Biomarkers           0   
1                                     Research Scientist           4   
2                    ASP.NET / C# / SQL 2005 / XML / XSD          32   
3                                            PHP / MySQL          32   
4                                      ASP / CRM / HRXML          40   
...                                                  ...         ...   
65338               Senior Director, Product Development     1760983   
65339                                          Associate       20167   
65340                    District Team Leader - Oklahoma     1761443   
65341  Director - Head of CDO Management, Asset Backe...      289179   
65342                                          Associate     1761459   

                                   org  
0                  Johnson and Johnson  
1      Sahlgrenska University Hospital  
2        <Software Consulting Company>  
3        <Software Consulting Company>  
4                  <Employment Agency>  
...                                ...  
65338        Ticketmaster - CitySearch  
65339                             AT&T  
65340                    Target Stores  
65341           Wells Fargo Securities  
65342    Cadwalader, Wickersham & Taft  

[256823 rows x 6 columns]

Again, this matches our Pandas query.

## Two Hops

Can we find employees who have a common employer?

Let's divide it into steps.

First, let's do a building block, namely the set of people, companies they work for, and the interconnecting `works` relationships:

```
people --> works_for --> company <-- works_for
[---- (persons_companies) -----]
```

In [34]:
# sharded_by_companies is the works_for edge, sharded by company
second_works_for = rename_shards(sharded_by_companies, {'_id': '_id2', 'role': 'role2'})

# We are currently sharded on company
person_works_company_works = \
    filter_shards(join_shards(persons_companies, second_works_for, 'company_id', 'company_id'),
           lambda row: row['_id'] != row['_id2']
    )
pd.concat(person_works_company_works)

_id family_name given_name  \
1         adagio-catalyst        Watt    Brunton   
2         adagio-catalyst        Watt    Brunton   
3         adagio-catalyst        Watt    Brunton   
4         adagio-catalyst        Watt    Brunton   
31            exact-novel        Morr  Barrymore   
...                   ...         ...        ...   
1703227  espressivo-media      Dundas    Brunton   
1703228  espressivo-media      Dundas    Brunton   
1703231   resultant-coach     Kincaid  Bertuccio   
1703232   resultant-coach     Kincaid  Bertuccio   
1703233   resultant-coach     Kincaid  Bertuccio   

                                                   title_x  company_id  \
1                    Senior Scientist, Oncology Biomarkers           0   
2                    Senior Scientist, Oncology Biomarkers           0   
3                    Senior Scientist, Oncology Biomarkers           0   
4                    Senior Scientist, Oncology Biomarkers           0   
31                     marketing and communication manager        1412   
...                                                    ...         ...   
1703227                                          Associate       20167   
1703228                                          Associate       20167   
1703231  Director - Head of CDO Management, Asset Backe...      289179   
1703232  Director - Head of CDO Management, Asset Backe...      289179   
1703233  Director - Head of CDO Management, Asset Backe...      289179   

                            org                    _id2  \
1           Johnson and Johnson          paper-scarfing   
2           Johnson and Johnson          paper-scarfing   
3           Johnson and Johnson              sharp-deck   
4           Johnson and Johnson               stern-bus   
31                      Sephora           fixed-triplet   
...                         ...                     ...   
1703227                    AT&T         overcast-jetsam   
1703228                    AT&T  associative-restaurant   
1703231  Wells Fargo Securities       endothermic-grape   
1703232  Wells Fargo Securities        spatial-strategy   
1703233  Wells Fargo Securities              icy-router   

                                                   title_y  
1        GOLD Associate (Global Operations Leadership D...  
2                                            Summer Intern  
3                        Professional Sales Representative  
4                                        Senior Consultant  
31                                                  Intern  
...                                                    ...  
1703227                               OI DSG Engineer, GT3  
1703228   Intern, Sales Compliance and Contract Management  
1703231                             Sr. ETL/Data Architect  
1703232                             Quantitative Associate  
1703233                                Software Consultant  

[4660010 rows x 8 columns]

Now we can join with another company of the

In [35]:
sharded_persons_2 = rename_shards(sharded_persons, {'_id': '_id2', 'given_name': 'given_name2', 'last_name': 'last_name2'})

result = join_shards(repartition(person_works_company_works, '_id2'), sharded_persons_2, '_id2', '_id2')
# Take the above and shard + join with sharded_persons
for table in result:
  display(table)

display(pd.concat(result))

_id family_name_x given_name  \
0             exact-novel          Morr  Barrymore   
1             exact-novel          Morr  Barrymore   
2             exact-novel          Morr  Barrymore   
3           massive-truss        Forbes    Jenkins   
4           massive-truss        Forbes    Jenkins   
...                   ...           ...        ...   
1268927  espressivo-media        Dundas    Brunton   
1268928  espressivo-media        Dundas    Brunton   
1268929  espressivo-media        Dundas    Brunton   
1268930  espressivo-media        Dundas    Brunton   
1268931   resultant-coach       Kincaid  Bertuccio   

                                                   title_x  company_id  \
0                      marketing and communication manager        1412   
1                      marketing and communication manager        1412   
2                      marketing and communication manager        1412   
3                  Research & Consulting Software Engineer        3512   
4                  Research & Consulting Software Engineer        3512   
...                                                    ...         ...   
1268927                                          Associate       20167   
1268928                                          Associate       20167   
1268929                                          Associate       20167   
1268930                                          Associate       20167   
1268931  Director - Head of CDO Management, Asset Backe...      289179   

                            org                      _id2  \
0                       Sephora  nondeterministic-fractal   
1                       Sephora                dark-throw   
2                       Sephora                dark-throw   
3                    LexisNexis                joint-plot   
4                    LexisNexis              tossed-queue   
...                         ...                       ...   
1268927                    AT&T           noisy-elevation   
1268928                    AT&T         tart-displacement   
1268929                    AT&T          primordial-story   
1268930                    AT&T             windy-parkway   
1268931  Wells Fargo Securities                icy-router   

                                        title_y family_name_y   given_name2  
0                VP Technical Services and CISO       Forsyth        Jeeves  
1           Talent Development Manager (People)       Macneil     Barrymore  
2                       Training Store Director       Macneil     Barrymore  
3                  Business Development Manager       Douglas     Bertuccio  
4                          Solutions Consultant        Brodie        Figaro  
...                                         ...           ...           ...  
1268927                        Staff Supervisor     Robertson       Cadbury  
1268928           Sr. Member of Technical Staff       Haldane        Figaro  
1268929  Manager Systems Design and Development    Sutherland  Passepartout  
1268930                      Technical Director        Gordon     Bertuccio  
1268931                     Software Consultant          Gunn      Merriman  

[1268932 rows x 10 columns]

_id family_name_x given_name  \
0             exact-novel          Morr  Barrymore   
1             exact-novel          Morr  Barrymore   
2             exact-novel          Morr  Barrymore   
3           massive-truss        Forbes    Jenkins   
4           massive-truss        Forbes    Jenkins   
...                   ...           ...        ...   
1326689  espressivo-media        Dundas    Brunton   
1326690  espressivo-media        Dundas    Brunton   
1326691  espressivo-media        Dundas    Brunton   
1326692  espressivo-media        Dundas    Brunton   
1326693  espressivo-media        Dundas    Brunton   

                                         title_x  company_id         org  \
0            marketing and communication manager        1412     Sephora   
1            marketing and communication manager        1412     Sephora   
2            marketing and communication manager        1412     Sephora   
3        Research & Consulting Software Engineer        3512  LexisNexis   
4        Research & Consulting Software Engineer        3512  LexisNexis   
...                                          ...         ...         ...   
1326689                                Associate       20167        AT&T   
1326690                                Associate       20167        AT&T   
1326691                                Associate       20167        AT&T   
1326692                                Associate       20167        AT&T   
1326693                                Associate       20167        AT&T   

                             _id2                                 title_y  \
0                 andante-account          Senior Product Manager, Social   
1        undecidable-differential                 Server Systems Engineer   
2                 concrete-galley  Senior Project Manager IT (consultant)   
3                    giant-bounce                       Student Associate   
4                   grouchy-ranch    Manager of Quality Assurance - Juris   
...                           ...                                     ...   
1326689           andante-stadium                  Vice President - Sales   
1326690           andante-stadium        Vice President - General Manager   
1326691             crazy-trigger                      Principal Engineer   
1326692              sunny-render            Support & Project Coodinator   
1326693            elegant-volume                      Network Specialist   

        family_name_y   given_name2  
0            Campbell      Merriman  
1              Murray     Simonides  
2              Fraser       Jenkins  
3            Hamilton     Simonides  
4              Fraser        Alfred  
...               ...           ...  
1326689    Macdougall     Bullimore  
1326690    Macdougall     Bullimore  
1326691       Barclay       Cadbury  
1326692          Mair       Barkley  
1326693      Hamilton  Passepartout  

[1326694 rows x 10 columns]

_id family_name_x given_name  \
0         adagio-catalyst          Watt    Brunton   
1         adagio-catalyst          Watt    Brunton   
2             exact-novel          Morr  Barrymore   
3             exact-novel          Morr  Barrymore   
4             exact-novel          Morr  Barrymore   
...                   ...           ...        ...   
1299889  espressivo-media        Dundas    Brunton   
1299890  espressivo-media        Dundas    Brunton   
1299891  espressivo-media        Dundas    Brunton   
1299892  espressivo-media        Dundas    Brunton   
1299893   resultant-coach       Kincaid  Bertuccio   

                                                   title_x  company_id  \
0                    Senior Scientist, Oncology Biomarkers           0   
1                    Senior Scientist, Oncology Biomarkers           0   
2                      marketing and communication manager        1412   
3                      marketing and communication manager        1412   
4                      marketing and communication manager        1412   
...                                                    ...         ...   
1299889                                          Associate       20167   
1299890                                          Associate       20167   
1299891                                          Associate       20167   
1299892                                          Associate       20167   
1299893  Director - Head of CDO Management, Asset Backe...      289179   

                            org              _id2  \
0           Johnson and Johnson        sharp-deck   
1           Johnson and Johnson         stern-bus   
2                       Sephora     fixed-triplet   
3                       Sephora  inverse-balinese   
4                       Sephora          one-helm   
...                         ...               ...   
1299889                    AT&T     congruent-app   
1299890                    AT&T     upbeat-bridge   
1299891                    AT&T     vertical-spot   
1299892                    AT&T      sleek-puffin   
1299893  Wells Fargo Securities  spatial-strategy   

                                                   title_y family_name_y  \
0                        Professional Sales Representative      Maclaren   
1                                        Senior Consultant      Buchanan   
2                                                   Intern       Malcolm   
3                                      Make-up Artist (PT)      Oliphant   
4                         Category Manager Europe Skincare      Sterling   
...                                                    ...           ...   
1299889             Information System Senior Staff Member      Grierson   
1299890                                     Senior Analyst           Hay   
1299891                                         Consultant    Montgomery   
1299892  Sr .Vice President, Network Development & Rese...       Burnett   
1299893                             Quantitative Associate       Riddell   

        given_name2  
0         Barrymore  
1           Barkley  
2            Figaro  
3         Bertuccio  
4         Leporello  
...             ...  
1299889   Simonides  
1299890   Belvedere  
1299891    Merriman  
1299892      Alfred  
1299893   Belvedere  

[1299894 rows x 10 columns]

_id family_name_x given_name  \
0         adagio-catalyst          Watt    Brunton   
1         adagio-catalyst          Watt    Brunton   
2             exact-novel          Morr  Barrymore   
3             exact-novel          Morr  Barrymore   
4           massive-truss        Forbes    Jenkins   
...                   ...           ...        ...   
1307399  espressivo-media        Dundas    Brunton   
1307400  espressivo-media        Dundas    Brunton   
1307401  espressivo-media        Dundas    Brunton   
1307402  espressivo-media        Dundas    Brunton   
1307403   resultant-coach       Kincaid  Bertuccio   

                                                   title_x  company_id  \
0                    Senior Scientist, Oncology Biomarkers           0   
1                    Senior Scientist, Oncology Biomarkers           0   
2                      marketing and communication manager        1412   
3                      marketing and communication manager        1412   
4                  Research & Consulting Software Engineer        3512   
...                                                    ...         ...   
1307399                                          Associate       20167   
1307400                                          Associate       20167   
1307401                                          Associate       20167   
1307402                                          Associate       20167   
1307403  Director - Head of CDO Management, Asset Backe...      289179   

                            org                    _id2  \
0           Johnson and Johnson          paper-scarfing   
1           Johnson and Johnson          paper-scarfing   
2                       Sephora          spicy-template   
3                       Sephora       joint-opportunity   
4                    LexisNexis      decidable-loveseat   
...                         ...                     ...   
1307399                    AT&T          stubborn-glass   
1307400                    AT&T           amber-dessert   
1307401                    AT&T         overcast-jetsam   
1307402                    AT&T  associative-restaurant   
1307403  Wells Fargo Securities       endothermic-grape   

                                                   title_y family_name_y  \
0        GOLD Associate (Global Operations Leadership D...    Montgomery   
1                                            Summer Intern    Montgomery   
2                                          Senior Designer       Kennedy   
3                                        Assistant Manager       Macneil   
4                                Human Resources Executive     Macmillan   
...                                                    ...           ...   
1307399                   Enterprise Application Developer       Chattan   
1307400                                         Contractor    Macpherson   
1307401                               OI DSG Engineer, GT3          Morr   
1307402   Intern, Sales Compliance and Contract Management       Elliott   
1307403                             Sr. ETL/Data Architect       Forsyth   

        given_name2  
0           Cadbury  
1           Cadbury  
2             Poole  
3           Cadbury  
4           Brunton  
...             ...  
1307399    Merriman  
1307400      Alfred  
1307401   Bullimore  
1307402     Brunton  
1307403   Simonides  

[1307404 rows x 10 columns]

_id family_name_x given_name  \
0             exact-novel          Morr  Barrymore   
1             exact-novel          Morr  Barrymore   
2             exact-novel          Morr  Barrymore   
3           massive-truss        Forbes    Jenkins   
4           massive-truss        Forbes    Jenkins   
...                   ...           ...        ...   
1307399  espressivo-media        Dundas    Brunton   
1307400  espressivo-media        Dundas    Brunton   
1307401  espressivo-media        Dundas    Brunton   
1307402  espressivo-media        Dundas    Brunton   
1307403   resultant-coach       Kincaid  Bertuccio   

                                                   title_x  company_id  \
0                      marketing and communication manager        1412   
1                      marketing and communication manager        1412   
2                      marketing and communication manager        1412   
3                  Research & Consulting Software Engineer        3512   
4                  Research & Consulting Software Engineer        3512   
...                                                    ...         ...   
1307399                                          Associate       20167   
1307400                                          Associate       20167   
1307401                                          Associate       20167   
1307402                                          Associate       20167   
1307403  Director - Head of CDO Management, Asset Backe...      289179   

                            org                      _id2  \
0                       Sephora  nondeterministic-fractal   
1                       Sephora                dark-throw   
2                       Sephora                dark-throw   
3                    LexisNexis                joint-plot   
4                    LexisNexis              tossed-queue   
...                         ...                       ...   
1307399                    AT&T            stubborn-glass   
1307400                    AT&T             amber-dessert   
1307401                    AT&T           overcast-jetsam   
1307402                    AT&T    associative-restaurant   
1307403  Wells Fargo Securities         endothermic-grape   

                                                  title_y family_name_y  \
0                          VP Technical Services and CISO       Forsyth   
1                     Talent Development Manager (People)       Macneil   
2                                 Training Store Director       Macneil   
3                            Business Development Manager       Douglas   
4                                    Solutions Consultant        Brodie   
...                                                   ...           ...   
1307399                  Enterprise Application Developer       Chattan   
1307400                                        Contractor    Macpherson   
1307401                              OI DSG Engineer, GT3          Morr   
1307402  Intern, Sales Compliance and Contract Management       Elliott   
1307403                            Sr. ETL/Data Architect       Forsyth   

        given_name2  
0            Jeeves  
1         Barrymore  
2         Barrymore  
3         Bertuccio  
4            Figaro  
...             ...  
1307399    Merriman  
1307400      Alfred  
1307401   Bullimore  
1307402     Brunton  
1307403   Simonides  

[5202924 rows x 10 columns]

## Exercise

Let's do this via a broadcast join.

In [36]:
# TODO
persons_2_df = persons_df.rename(columns={
    '_id': '_id2',
    'given_name': 'given_name2',
    'last_name': 'last_name2'
})
result = join_shard_with_broadcast(person_works_company_works,
    persons_2_df,
    '_id2',
    '_id2'
)

display(pd.concat(result))

_id family_name_x given_name  \
0         adagio-catalyst          Watt    Brunton   
1         adagio-catalyst          Watt    Brunton   
2         adagio-catalyst          Watt    Brunton   
3         adagio-catalyst          Watt    Brunton   
4             exact-novel          Morr  Barrymore   
...                   ...           ...        ...   
1786613  espressivo-media        Dundas    Brunton   
1786614  espressivo-media        Dundas    Brunton   
1786615   resultant-coach       Kincaid  Bertuccio   
1786616   resultant-coach       Kincaid  Bertuccio   
1786617   resultant-coach       Kincaid  Bertuccio   

                                                   title_x  company_id  \
0                    Senior Scientist, Oncology Biomarkers           0   
1                    Senior Scientist, Oncology Biomarkers           0   
2                    Senior Scientist, Oncology Biomarkers           0   
3                    Senior Scientist, Oncology Biomarkers           0   
4                      marketing and communication manager        1412   
...                                                    ...         ...   
1786613                                          Associate       20167   
1786614                                          Associate       20167   
1786615  Director - Head of CDO Management, Asset Backe...      289179   
1786616  Director - Head of CDO Management, Asset Backe...      289179   
1786617  Director - Head of CDO Management, Asset Backe...      289179   

                            org                    _id2  \
0           Johnson and Johnson          paper-scarfing   
1           Johnson and Johnson          paper-scarfing   
2           Johnson and Johnson              sharp-deck   
3           Johnson and Johnson               stern-bus   
4                       Sephora           fixed-triplet   
...                         ...                     ...   
1786613                    AT&T         overcast-jetsam   
1786614                    AT&T  associative-restaurant   
1786615  Wells Fargo Securities       endothermic-grape   
1786616  Wells Fargo Securities        spatial-strategy   
1786617  Wells Fargo Securities              icy-router   

                                                   title_y family_name_y  \
0        GOLD Associate (Global Operations Leadership D...    Montgomery   
1                                            Summer Intern    Montgomery   
2                        Professional Sales Representative      Maclaren   
3                                        Senior Consultant      Buchanan   
4                                                   Intern       Malcolm   
...                                                    ...           ...   
1786613                               OI DSG Engineer, GT3          Morr   
1786614   Intern, Sales Compliance and Contract Management       Elliott   
1786615                             Sr. ETL/Data Architect       Forsyth   
1786616                             Quantitative Associate       Riddell   
1786617                                Software Consultant          Gunn   

        given_name2  
0           Cadbury  
1           Cadbury  
2         Barrymore  
3           Barkley  
4            Figaro  
...             ...  
1786613   Bullimore  
1786614     Brunton  
1786615   Simonides  
1786616   Belvedere  
1786617    Merriman  

[5202924 rows x 10 columns]

In [37]:
grader.grade('sharded_net_result', len(pd.concat(result)))

Correct! You earned 1/1 points. You are a star!

Your submission has been successfully recorded in the gradebook.
